packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -411,32 +412,66 @@ class TextPainter {
     return _isUtf16Surrogate(prevCodeUnit) ? offset - 2 : offset - 1;
   }
 
+  // Unicode value for a zero width joiner character.
+  static const int _zwjUtf16 = 0x200d;
+
+  // TODO(garyq): Use actual extended grapheme cluster length instead of
+  // an increasing cluster length amount to achieve deterministic performance.
   Offset _getOffsetFromUpstream(int offset, Rect caretPrototype) {
-    final int prevCodeUnit = _text.codeUnitAt(offset - 1);
+    final int prevCodeUnit = _text.codeUnitAt(max(0, offset - 1));
     if (prevCodeUnit == null)
       return null;
-    final int prevRuneOffset = _isUtf16Surrogate(prevCodeUnit) ? offset - 2 : offset - 1;
-    final List<TextBox> boxes = _paragraph.getBoxesForRange(prevRuneOffset, offset);
-    if (boxes.isEmpty)
-      return null;
-    final TextBox box = boxes[0];
-    final double caretEnd = box.end;
-    final double dx = box.direction == TextDirection.rtl ? caretEnd - caretPrototype.width : caretEnd;
-    return Offset(dx, box.top);
+    // Check for multi-code-unit glyphs such as emojis or zero width joiner
+    final bool needsSearch = _isUtf16Surrogate(prevCodeUnit) || _text.codeUnitAt(offset) == _zwjUtf16;
+    int graphemeClusterLength = needsSearch ? 2 : 1;
+    List<TextBox> boxes = <TextBox>[];
+    while (boxes.isEmpty && _text.text != null) {
+      final int prevRuneOffset = offset - graphemeClusterLength;
+      boxes = _paragraph.getBoxesForRange(prevRuneOffset, offset);
+      if (boxes.isEmpty) {
```

Comment:
Be liberal with comments and document any idiosyncrasies you're depending on (e.g. it returning empty if the range doesn't encompass an entire cluster).